### ISSUE TITLE
import: importing should use manual splitting with sticky bits

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#37697]
Thanks to @jeffrey-xiao's work, one can now use `alter table split at` while the merge queue is enabled. The split range will have a sticy bit.
Imports (and restores) should use this mechanism for generating the splits they need instead of doing whatever they do today. I don't know exactly what they do but I believe it involves a funky mechanism of gossiping some information so that the merge queue doesn't undo the splits that they've somehow made. And then the restore should "unsplit" everything once it's done (which unsplit mechanism I believe is also on the verge of existing).
Right, @dt?

@jeffrey-xiao , are you interested in working on this?

cc @nvanbenschoten @ajwerner

**Comment by tbg** [comment]
Note that it may be more trouble than it's worth to rip out the gossip code and any other workarounds before 20.1 is the next release because 19.2 will have to interop with 19.1 in a mixed cluster. I wouldn't suggest anyone to run a restore during a version upgrade, but I also don't think it's worth running into this kind of problem (see also roachtests) eagerly. That said, if the new cluster version is active we should use manual splits to weed out any problems.

**Comment by dt** [comment]
I missed the initial discussion of sticky bits. Looking at #37728 now, I have a couple questions/concerns.

TL;DR: I think the setting is dangerously coarse. I think we should consider the split itself reflecting how long it thinks it should live, or, if not, `UNSPLIT AT` should be smarter.

So it seems to me that there is difference between the general case of "I want this table split at this point" and "I'm pre-splitting this table before ingestion" (both when done automatically by IMPORT or RESTORE or done by hand before bulk-loading something).

Putting the split time in the sticky bit erases this distinction. Using a cluster setting to ignore all sticky bits older than X treats a split that one team added to their table at a particular hot key exactly the same as the splits another team added to a different table before a big bulk-load. I appreciate the simplicity, but fear that the cluster-wide setting as is is so coarse as to be a somewhat scary potential foot-gun in production clusters.

What if, instead of putting the current time in the sticky bit, we put the intended expiration time? That could allow for saying `SPLIT AT X EXPIRES '12h'` for example, which makes explicit when _that split_ can be removed, avoiding some of the potential messy interactions across tables being used differently that the global setting introduces.

Concrete strawman: AdminSplit takes a expiration timestamp. 0 means not sticky/it can expire whenever you want, some non-zero time is when its stickiness expires, max int means never expires. Extend `SPLIT AT` syntax with an optional `EXPIRES duration` modifier. Bulk-IO jobs can pick a TTL that should give them enough time to start importing into the split range, after which normal merge logic is fine.

However, if we don't like that and instead want bulk-IO jobs to a) make splits that have no expiration and then b) iterate all the ranges in their ingest span and clear the sticky bit then I think that approach should also be the approach used to clear manual pre-splits, rather than the setting and its potential for unintended consequences. If we have to build something that gets the ranges for a span and unsplits each for IMPORT and RESTORE, we might as well use it to make a `ALTER TABLE UNSPLIT ALL` too.

**Comment by andreimatei** [comment]
I agree with @dt in that I don't see much of a point in the existence of this cluster setting.
I definitely was not suggesting that we rely on it for unsplitting after an import is done.

I think having (and even requiring) an expiration that's set on the sticky bit itself would be preferable to a catch-all setting; I think no sticky bit should persist forever, but I think we need a less blunt instrument than a cluster-wide setting.
Then, I think indeed we should have an `ALTER TABLE UNSPLIT ALL`; I think that makes sense. This could be the preferred mechanism for clearing splits done before a manual data loading operation. And imports and restores could use this too, but see below.

For imports and/or restores, @dt was telling me that (at least for one of them) the split points are created just before ingesting the data for one range, so a TTL of a few seconds should be enough. For the cases where a tiny TTL is supposed to be enough, I think should do exactly that.
When a short TTL is not enough, perhaps we can remove a split point as the restoring happens, once both sides of the split have been filled in. To avoid the extra step at the end that iterates through all the table's ranges.

**Comment by nvanbenschoten** [comment]
I agree with a lot of what @dt said as well. 

Switching the timestamp in the sticky bit from the split timestamp to an expiration timestamp makes a lot of sense because it allows for more fine-grained ephemeral split policies. The concrete strawman LGTM.

An `ALTER TABLE UNSPLIT ALL` statement also makes sense.
